### ISSUE TITLE
Don't overwrite existing tags when tagging spot instances

### ISSUE THREAD

**Issue by cristim** [cristim/autospotting#185]
# Github issue #

## Issue type ##

- Bug Report

## Build number #

Current as of 9 Dec 2017

## Configuration ##

Launch configuration that contains-user-data script that updates instance tags at start (sets Name to match the Instance ID).  

## Summary ##

Reported by @sc-chad on Gitter:

> If you change or update the Tag:Name (we do this to uniquely identify instances by adding the instance ID to the name), the lambda function seems to revert it back to the original name.

## Steps to reproduce ##

Have AutoSpotting start a spot instance on a group with relatively large grace period, and propagate the a tag set to a given arbitrary value to the instances.

Before the instance is added to the group, alter the tag to match the instance ID or any other arbitrary value.

## Expected results ##

The tag is reverted to the original value set on the group, instead of keeping the updated value.

## Actual results ##
The tag value should not be changed if existing.

**Comment by sc-chad** [comment]
This is my last hurdle before I can roll this out across our infrastructure.

@cristim I'm guessing I need to start around here: https://github.com/cristim/autospotting/blob/687c852039cc2fea0b88b6b7d931626367a73094/core/spot_instance_request.go#L72

If someone else wants to take a whack at this I'd be very greatful!

**Comment by cristim** [comment]
@sc-chad no, that's for something else.

You want to play around https://github.com/cristim/autospotting/blob/687c852039cc2fea0b88b6b7d931626367a73094/core/instance.go#L267 and/or https://github.com/cristim/autospotting/blob/687c852039cc2fea0b88b6b7d931626367a73094/core/autoscaling.go#L272 and/or https://github.com/cristim/autospotting/blob/687c852039cc2fea0b88b6b7d931626367a73094/core/spot_instance_request.go#L23:31

Basically what you should to is simply skip all the tags already set on the instance being tagged.

**Comment by xlr-8** [comment]
@sc-chad I've looking a bit at the code and I couldn't (yet?) find a rule that would "overwrite" tags meant to be on the instance.

My understanding of 

> Have AutoSpotting start a spot instance on a group with relatively large grace period, and propagate the a tag set to a given arbitrary value to the instances.
> 
> Before the instance is added to the group, alter the tag to match the instance ID or any other arbitrary value.

Is that you have for example a tag "deploying" being set to "in progress" by the ASG (rule to propagate set to yes), and you have your user-data modifying this value to "done" while being started/deployed.

Your issue is then that this tag is reverted to "test" instead of being to "done" as set by the instance. Is this correct?

**Comment by sc-chad** [comment]
@xlr-8 Actually after doing a bunch of testing on Friday.  I believe the "reset" of tags is actually when it is attaching it to the autoscaling group.

If I update the tags after it's a member of the ASG, it never gets overwrote.

So this issue is actually very minor since all I need to do is update it on that event, but it would probably be nice if it didn't do that either if possible.

On your question about using user-data.  Basically I have a bash script that is ran when the instance starts that:

- checks the name tag if it contains its instance id
- if the name tag doesn't it contain the instance id, it updates it and the hostname

The core of my issue is if I had a tag:Name set to `foo` on the ASG and to propagate to all instances.  I would set the instance tag:Name to `foo-12345` on instance start.  At some point in the life of the spot instance being started and the old one terminated, the tag:Name was reverted back to `foo`.

I believe this is when it is attached to the ASG, but I solved it now for my case by periodically checking if the tag:Name isn't updated and rerunning the update script.

**Comment by cristim** [comment]
That's true, the instances are tagged twice: once immediately after launching, in order to support setups depending on tags for bootstrapping, and once again when they are added to the group, in order to bring them in sync with whatever tags were set on the group at the time.

**Comment by xlr-8** [comment]
As underlined by @cristim a viable option would be to start using the [RunInstances](http://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.RunInstances) which would allow to directly tag the spot instances when started. I'll take a look at it

**Comment by cristim** [comment]
@sc-chad is this still an issue that blocks you from using Autospotting?

**Comment by sc-chad** [comment]
Nah I worked around it by running a script that checks if the tags have been set.  Thanks guys.

**Comment by cristim** [comment]
This should no longer be a problem. The tags are only set once, when starting the new spot instances. Any further changes by the user-data scripts should no longer be reverted.

**Comment by sc-chad** [comment]
Awesome thanks!
